src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -983,240 +826,66 @@ public SolutionState WithProjectMetadataReferences(ProjectId projectId, IReadOnl
-        {
-            return AddDocumentsToMultipleProjects(documentInfos,
-                (documentInfo, project) => project.CreateDocument(documentInfo, project.ParseOptions, new LoadTextOptions(project.ChecksumAlgorithm)),
-                (oldProject, documents) => (oldProject.AddDocuments(documents), new CompilationAndGeneratorDriverTranslationAction.AddDocumentsAction(documents)));
-        }
-
-        /// <summary>
-        /// Core helper that takes a set of <see cref="DocumentInfo" />s and does the application of the appropriate documents to each project.
-        /// </summary>
-        /// <param name="documentInfos">The set of documents to add.</param>
-        /// <param name="addDocumentsToProjectState">Returns the new <see cref="ProjectState"/> with the documents added, and the <see cref="CompilationAndGeneratorDriverTranslationAction"/> needed as well.</param>
-        /// <returns></returns>
-        private SolutionState AddDocumentsToMultipleProjects<T>(
-            ImmutableArray<DocumentInfo> documentInfos,
-            Func<DocumentInfo, ProjectState, T> createDocumentState,
-            Func<ProjectState, ImmutableArray<T>, (ProjectState newState, CompilationAndGeneratorDriverTranslationAction translationAction)> addDocumentsToProjectState)
-            where T : TextDocumentState
-        {
-            if (documentInfos.IsDefault)
-            {
-                throw new ArgumentNullException(nameof(documentInfos));
-            }
-
-            if (documentInfos.IsEmpty)
-            {
-                return this;
-            }
-
-            // The documents might be contributing to multiple different projects; split them by project and then we'll process
-            // project-at-a-time.
-            var documentInfosByProjectId = documentInfos.ToLookup(d => d.Id.ProjectId);
-
-            var newSolutionState = this;
-
-            foreach (var documentInfosInProject in documentInfosByProjectId)
-            {
-                CheckContainsProject(documentInfosInProject.Key);
-                var oldProjectState = this.GetProjectState(documentInfosInProject.Key)!;
-
-                var newDocumentStatesForProjectBuilder = ArrayBuilder<T>.GetInstance();
-
-                foreach (var documentInfo in documentInfosInProject)
-                {
-                    newDocumentStatesForProjectBuilder.Add(createDocumentState(documentInfo, oldProjectState));
-                }
-
-                var newDocumentStatesForProject = newDocumentStatesForProjectBuilder.ToImmutableAndFree();
-
-                var (newProjectState, compilationTranslationAction) = addDocumentsToProjectState(oldProjectState, newDocumentStatesForProject);
-
-                newSolutionState = newSolutionState.ForkProject(newProjectState,
-                    compilationTranslationAction,
-                    newFilePathToDocumentIdsMap: CreateFilePathToDocumentIdsMapWithAddedDocuments(newDocumentStatesForProject));
```

Comment:
@jasonmalinowski A few important things here:

1. when this method moves to CompilationState it needs to both make a new SolutionState anda  new CompilationState.
2. For the new SOlutionState it still passes in CreateFilePathToDocumentIdsMapWithAddedDocuments.  Note though that CreateFilePathToDocumentIdsMapWithAddedDocuments is called on `this`, not on `newSolutionState`.  Not sure if that's intentional or not.  I have preserved that logic in the callsite.  It will call CreateFilePathToDocumentIdsMapWithAddedDocuments on the *original* solution state, not the intermediary one it is building.